plutus-tx/src/PlutusTx/Ratio.hs

```diff
@@ -4,80 +4,362 @@
+  Rational n d >= Rational n' d' = (n P.* d') P.>= (n' P.* d)
+  {-# INLINABLE (<) #-}
+  Rational n d < Rational n' d' = (n P.* d') P.< (n' P.* d)
+  {-# INLINABLE (>) #-}
+  Rational n d > Rational n' d' = (n P.* d') P.> (n' P.* d)
+
+instance Ord Rational where
+  compare (Rational n d) (Rational n' d') = compare (n * d') (n' * d)
+  Rational n d <= Rational n' d' = (n * d') <= (n' * d)
+  Rational n d >= Rational n' d' = (n * d') >= (n' * d)
+  Rational n d < Rational n' d' = (n * d') < (n' * d)
+  Rational n d > Rational n' d' = (n * d') > (n' * d)
+
+instance P.AdditiveSemigroup Rational where
+  {-# INLINABLE (+) #-}
+  Rational n d + Rational n' d' =
+    let newNum = (n P.* d') P.+ (n' P.* d)
+        newDen = d P.* d'
+        gcd' = euclid newNum newDen
+     in Rational (newNum `Builtins.quotientInteger` gcd')
+                 (newDen `Builtins.quotientInteger` gcd')
+
+instance P.AdditiveMonoid Rational where
+  {-# INLINABLE zero #-}
+  zero = Rational P.zero P.one
+
+instance P.AdditiveGroup Rational where
+  {-# INLINABLE (-) #-}
+  Rational n d - Rational n' d' =
+    let newNum = (n P.* d') P.- (n' P.* d)
+        newDen = d P.* d'
+        gcd' = euclid newNum newDen
+     in Rational (newNum `Builtins.quotientInteger` gcd')
+                 (newDen `Builtins.quotientInteger` gcd')
+
+instance P.MultiplicativeSemigroup Rational where
+  {-# INLINABLE (*) #-}
+  Rational n d * Rational n' d' =
+    let newNum = n P.* n'
+        newDen = d P.* d'
+        gcd' = euclid newNum newDen
+     in Rational (newNum `Builtins.quotientInteger` gcd')
+                 (newDen `Builtins.quotientInteger` gcd')
+
+instance P.MultiplicativeMonoid Rational where
+  {-# INLINABLE one #-}
+  one = Rational P.one P.one
+
+instance P.Module Integer Rational where
+  {-# INLINABLE scale #-}
+  scale i (Rational n d) = let newNum = i P.* n
+                               gcd' = euclid newNum d in
+    Rational (newNum `Builtins.quotientInteger` gcd')
+             (d `Builtins.quotientInteger` gcd')
+
+instance P.ToData Rational where
+  {-# INLINABLE toBuiltinData #-}
+  toBuiltinData (Rational n d) = P.toBuiltinData (n, d)
+
+-- These instances ensure that the following invariants don't break:
+--
+-- 1. The denominator is greater than 0; and
+-- 2. The numerator and denominator are coprime.
+--
+-- For invariant 1, fromBuiltinData yields Nothing on violation, while
+-- unsafeFromData calls error. Invariant 2 is kept maintained by use of
+-- unsafeRatio.
+
+instance P.FromData Rational where
+  {-# INLINABLE fromBuiltinData #-}
+  fromBuiltinData dat = case P.fromBuiltinData dat of
+    P.Nothing -> P.Nothing
+    P.Just (n, d) -> if d P.== P.zero
+                     then Builtins.error ()
+                     else P.Just (unsafeRatio n d)
+
+instance P.UnsafeFromData Rational where
+  {-# INLINABLE unsafeFromBuiltinData #-}
+  unsafeFromBuiltinData dat = case P.unsafeFromBuiltinData dat of
+    (n, d) -> if d P.== P.zero
```

Comment:
This one I think really could just be `unsafeRatio`.

Reply:
That's a very good catch - thank you for sticking to me about this! As a side effect, the new version of `fromBuiltinData` has an on-chain size that's 8 bytes smaller: not much, but every little helps.